source/extensions/listener_managers/listener_manager/fcds_impl.cc

```diff
@@ -0,0 +1,227 @@
+        {});
+  } else {
+      fcds_subscription_ = cluster_manager_.subscriptionFactory().collectionSubscriptionFromUrl(
+        fcds_resource_locator, fcds.config_source(), resource_name, *scope_, *this, resource_decoder_);
+  }
+}
+
+FcdsApi::~FcdsApi() {
+  // If we get destroyed during initialization, make sure we signal that we "initialized".
+  local_init_target_.ready();
+}
+
+void FcdsApi::initialize() {
+  local_init_manager_.add(local_init_target_);
+  init_manager_.add(parent_init_target_);
+}
+
+void FcdsApi::onConfigUpdate(const std::vector<Envoy::Config::DecodedResourceRef>& resources,
+                              const std::string& version_info) {
+  if (!validateUpdateSize(resources.size())) {
+    return;
+  }
+
+  // Construct the list of filter chains to be removed
+  std::unordered_set<std::string> deleted_filter_chains;
+  auto filter_chains_message_by_name = filter_chain_manager_.getExistingFilterChainMessages();
+  auto iter = filter_chains_message_by_name.begin();
+  while (iter != filter_chains_message_by_name.end()) {
+    deleted_filter_chains.insert(iter->first);
+    ++iter;
+  }
+
+  // Remove the list of FCs in the incoming FCDS config from the above list
+  for (const auto& filter_chain : resources) {
+    deleted_filter_chains.erase(filter_chain.get().name());
+  }
+
+  Protobuf::RepeatedPtrField<std::string> to_be_deleted_fc_list;
+  for (const auto& filter_chain : deleted_filter_chains) {
+    *to_be_deleted_fc_list.Add() = filter_chain;
+  }
+
+  // Delete the old fcs and add the new fcs
+  onConfigUpdate(resources, to_be_deleted_fc_list, version_info);
+}
+
+void FcdsApi::onConfigUpdate(const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+                              const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                              const std::string&) {
+
+  std::vector<const envoy::config::listener::v3::FilterChain*> removed_filter_chains;
+  std::vector<const envoy::config::listener::v3::FilterChain*> new_filter_chains;
+  std::string message;
+
+  TRY_ASSERT_MAIN_THREAD 
+    processDeletedFilterChains(removed_resources, removed_filter_chains);
+    processAddedOrUpdatedFilterChains(added_resources, removed_filter_chains, new_filter_chains);
+
+    // Finally invoke the deletion call
+    filter_chain_manager_.removeFilterChains(removed_filter_chains);
+
+    // Finally invoke the addition call
+    absl::Span<const envoy::config::listener::v3::FilterChain* const> filter_chain_span = new_filter_chains;
+    filter_chain_manager_.addFilterChains(filter_chain_span, nullptr, *fc_builder_,
+                                        filter_chain_manager_, true);
+
+    // Drain the filter chains which were deleted
+    filter_chain_manager_.startDrainingSequenceForListenerFilterChains();
+  END_TRY
+  catch (const EnvoyException& e)
+  {
+    absl::StrAppend(&message, e.what(), "\n");
```

Comment:
![fcds_comment_1](https://user-images.githubusercontent.com/8407970/217911045-4ba99ace-dc57-4b57-9a27-de713385c200.jpg)

Reply:
I think that it is not only "system" exceptions that may occur.
For example, it seems that `addFilterChains` may throw an exception. If that's the case, any chains removed by `removeFilterChains()` will disappear, and we end up in a wrong state.

Reply:
Rollback for exception cases is included in the latest commit.